docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,223 @@
+
+> You can find an example of a page model in the [testcafe-examples](https://github.com/DevExpress/testcafe-examples/tree/master/examples/use-page-model) repository.
+
+* Use the `.testcaferc.json` configuration file to fine-tune your tests, which can be useful in CI/CD systems. This file must be located in the root directory of the project. For more information, read the [Configuration File](../../reference/configuration-file.md) article.
+
+* Define one `fixture` in every test file. While it's technically possible to define multiple, this is not recommended.
+
+* TestCafe tests are purely functional. As such, they should not be concerned about the implementation details and it is best to isolate them from production code. Keep your test files in a separate directory. You can name this directory appropriately (for instance, `tests`).
+
+* In your test folder, group your tests into subfolders. These groups cover different subsystems of your application with tests.
+
+* Any reused data (for example, large sets of expected values or forms inputs) is better stored in a directory of its own. Consider a descriptive folder name (for instance, `data`).
+
+With all the suggestions applied, your project's file structure might look like this:
+
+```sh
+.
+├── .testcaferc.json
+└── tests
+    ├── |- test_group1/
+    │   └── |-test1.js
+    │       └── |-test2.js
+    ├── |- test_group2/
+    │   └── |-test1.js
+    │       └── |-test2.js
+    ├── |- page_model/
+    │   └── |- file1.js
+    │       └── |- file2.js
+    ├── |- helpers/
+    │   └── |- helper1.js
+    │       └── |- helper2.js
+    └── |-data
+```
+
+## Cleanup With beforeEach / State Management
+
+State management is an integral and important part of web testing. When your tests run, there are inevitably leftovers- database or local storage records, cache, or cookies. Your tests might require setup to run - like adding records to the database.
+
+One common strategy is to set up with the previous test's `after` and `afterEach` hooks. While good for cleanup, these hooks create mutual dependence between your tests when used to set up for the following test. The success rate of a test is then influenced by a preceding one, which is not desirable.
+
+Plus, if the setup with `before` or `beforeEach` is unsuccessful, the associated test does not run, which saves you time. If your setup with `after` and `afterEach` yields an error, the following test runs and probably fails due to the unsuccessful setup.
+
+To fulfill your test's state prerequisites, use the `before` and `beforeEach` hooks. Clean up with the `after` and `afterEach` hooks.
+
+Use the `test.before` and `test.after` hooks to set the state that an individual test requires. Use the `fixture.beforeEach` and `fixture.afterEach` to set common state that is required across the board.
+
+## E2E Tests Limitations
```

Comment:
This section is really good, I'm just concerned it is lost here.

Can we move it up? Probably it won't hurt to put it in the beginning (it's short enough).

And maybe *Scope* rather than *Limitation*?

Reply:
moved it to the top